### ISSUE TITLE
Serialization help

### ISSUE THREAD

**Issue by RecursiveVision** [LoneGazebo/Community-Patch-DLL#7296]
@ilteroi Hoping you could help me with this.

I finally got fed up with the overcomplicated diplo AI memory system, especially since it's giving me issues. 

I looked at CvMinorCivAI.h and noticed that all the arrays (including one multidimensional array, NumEnemyUnitsLeftToKillByMajor[MAX_MAJOR_CIVS][MAX_MAJOR_CIVS]) seemed to work just fine without this overcomplication, so I replaced every single variable's format to match this, including the 2D arrays.

The pushed branch is here: https://github.com/LoneGazebo/Community-Patch-DLL/commit/fbfb2338f64e166b3f28bab58bc983671af9a6d4

Serialization is now a lot easier on the eyes (and my sanity). It builds okay, and I suspect it will work once I'm able to test it.

However, I'm unsure as to whether the memory variables that are structures can be serialized properly - I think they may need additional work. There are two such variables:

```
DiploLogData m_aaDiploStatementsLog[MAX_MAJOR_CIVS][MAX_DIPLO_LOG_STATEMENTS];
DeclarationLogData m_aDeclarationsLog[MAX_DIPLO_LOG_STATEMENTS];
```
The structures are as follows:
```
struct DiploLogData
{
	DiploStatementTypes m_eDiploLogStatement;
	int m_iTurn;
};

struct DeclarationLogData
{
	PublicDeclarationTypes m_eDeclaration;
	int m_iData1;
	int m_iData2;
	PlayerTypes m_eMustHaveMetPlayer;
	bool m_bActive;
	int m_iTurn;
};
```
They reset as follows:
```
	for (int iI = 0; iI < MAX_MAJOR_CIVS; iI++)
	{
		for (int iJ = 0; iJ < MAX_DIPLO_LOG_STATEMENTS; iJ++)
		{
			m_aaDiploStatementsLog[iI][iJ].m_eDiploLogStatement = NO_DIPLO_STATEMENT_TYPE;
			m_aaDiploStatementsLog[iI][iJ].m_iTurn = -1;
		}
	}
	for (int iI = 0; iI < MAX_DIPLO_LOG_STATEMENTS; iI++)
	{
		m_aDeclarationsLog[iI].m_eDeclaration = NO_PUBLIC_DECLARATION_TYPE;
		m_aDeclarationsLog[iI].m_iTurn = -1;
	}
```
I suspect I can't simply do what I did for all the other variables, which is:
```
kStream >> m_aaDiploStatementsLog;
kStream >> m_aDeclarationsLog;

kStream << m_aaDiploStatementsLog;
kStream << m_aDeclarationsLog;
```

And have it remember everything correctly, even though it doesn't create any build errors. 

Do you know how I would correctly serialize this?

**Comment by ilteroi** [comment]
there are in fact generic templated serialization methods for containers, so as long as you have defined an operator << and operator >> for the structs, making the process relatively painless.

however, you are using raw C-style arrays right now, so you need to use an ArrayWrapper. alternatively you can switch to stl vectors.

**Comment by RecursiveVision** [comment]
@ilteroi Is the ArrayWrapper only required for containers like this, or do multidimensional arrays e.g. m_aaeCoopWarState[MAX_MAJOR_CIVS][MAX_MAJOR_CIVS] also require an ArrayWrapper?

**Comment by ilteroi** [comment]
well the ArrayWrapper is only 1D so the code you pasted seems ok? what's the error message?

**Comment by RecursiveVision** [comment]
![0d200e59cda9c265bacee0620d0ff862](https://user-images.githubusercontent.com/66801010/102339443-29c4af00-3f63-11eb-85cb-aff72ac62f33.png)

**Comment by RecursiveVision** [comment]
![Screenshot (8270)](https://user-images.githubusercontent.com/66801010/102339547-5082e580-3f63-11eb-82e8-d002c014f5c7.png)

**Comment by RecursiveVision** [comment]
No issue with the Read code, just the Write...weird how copypasted code is suddenly not working.

**Comment by RecursiveVision** [comment]
@ilteroi I pushed my latest code (with the problem variables commented out) here: 655d227

**Comment by ilteroi** [comment]
yeah, template pitfalls ... you need to provide a non-const pointer to the elements for initialization or maybe use ArrayWrapper<const x>

**Comment by RecursiveVision** [comment]
@ilteroi ArrayWrapper<const DiploLogData> fails to build: "you must implement a serializer for this operator"

How would I use a non-const pointer?

**Comment by RecursiveVision** [comment]
With the exception of these two containers, all other variables are simply arrays.

Either in 1D form:
`char m_aeMajorCivApproach[MAX_MAJOR_CIVS]`
or 2D form:
 `int m_aaiApproachValues[MAX_MAJOR_CIVS][NUM_MAJOR_CIV_APPROACHES]`

Will a simple kStream >> and << call work for these arrays?

**Comment by ilteroi** [comment]
ripple effects. redefine operator << to work with a const struct. or make the write() method non-const.

**Comment by RecursiveVision** [comment]
Making the write method non-const seems easiest, searching through the files show some other serialization is done that way. Will give it a try.

**Comment by RecursiveVision** [comment]
Well, it builds okay. Now the fun part: testing it!

Thanks for the help! :)

**Comment by RecursiveVision** [comment]
@LoneGazebo @ilteroi IT IS DONE! THE NEW SERIALIZATION WORKS! THE AWFUL DIPLOMACY MEMORY MANAGEMENT IS NO MORE!

I tested both 1D arrays (Declaration of Friendship) and 2D arrays (coop wars) using the debug mode I added. Worked fine, no issues with the new serialization or saving/loading. I didn't touch the code for containers (minus making :Write() non-const) as I don't want to mess with that, but aside from that it should be a lot more sane and easy to work with from now on.
